DynamicTablesPkg/Library/Acpi/RiscV/AcpiSsdtPlicAplicLib/SsdtPlicAplicGenerator.c

```diff
@@ -0,0 +1,588 @@
+#include <Library/AcpiHelperLib.h>
+#include <Library/TableHelperLib.h>
+#include <Library/AmlLib/AmlLib.h>
+#include <Protocol/ConfigurationManagerProtocol.h>
+
+// _SB scope of the AML namespace.
+#define SB_SCOPE  "\\_SB_"
+
+/** RISC-V SSDT PLIC/APLIC namespace device Generator.
+
+Requirements:
+  The following Configuration Manager Object(s) are used by
+  this Generator:
+  - ERiscVObjAplicInfo
+  - ERiscVObjPlicInfo
+*/
+
+/** This macro expands to a function that retrieves the APLIC
+    Information from the Configuration Manager.
+*/
+GET_OBJECT_LIST (
+  EObjNameSpaceRiscV,
+  ERiscVObjAplicInfo,
+  CM_RISCV_APLIC_INFO
+  );
+
+/** This macro expands to a function that retrieves the PLIC
+    Information from the Configuration Manager.
+*/
+
+GET_OBJECT_LIST (
+  EObjNameSpaceRiscV,
+  ERiscVObjPlicInfo,
+  CM_RISCV_PLIC_INFO
+  );
+
+/** Generate a PLIC/APLIC device.
+
+  @param [in]       Generator       The SSDT PlicAplic generator.
+  @param [in]       CfgMgrProtocol  Pointer to the Configuration Manager
+                                    Protocol interface.
+  @param [in]       BaseAddress     MMIO base of the PLIC/APLIC device.
+  @param [in]       Size            MMIO size of the PLIC/APLIC device.
+  @param [in]       GsiBase         GSI Base of the PLIC/APLIC device.
+  @param [in]       Uid             Unique Id of the PlicAplic device.
+  @param [in]       Hid             HID/CID of the interrupt controller device.
+  @param [in, out]  RootNode        RootNode of the AML tree to populate.
+
+  @retval EFI_SUCCESS               The function completed successfully.
+  @retval EFI_INVALID_PARAMETER     Invalid parameter.
+  @retval EFI_OUT_OF_RESOURCES      Could not allocate memory.
+**/
+STATIC
+EFI_STATUS
+EFIAPI
+GeneratePlicAplicDevice (
+  IN      CONST ACPI_TABLE_GENERATOR                          *Generator,
+  IN      CONST EDKII_CONFIGURATION_MANAGER_PROTOCOL  *CONST  CfgMgrProtocol,
+  IN            PLIC_APLIC_COMMON_INFO                        *PlicAplicInfo,
+  IN            CHAR8                                         *Hid,
+  IN            UINT32                                        Uid,
+  IN  OUT       AML_ROOT_NODE_HANDLE                          *RootNode
+  )
+{
+  AML_OBJECT_NODE_HANDLE  ScopeNode;
+  AML_OBJECT_NODE_HANDLE  IcNode;
+  AML_OBJECT_NODE_HANDLE  CrsNode;
+  EFI_STATUS              Status;
+  CHAR8                   AslName[AML_NAME_SEG_SIZE + 1];
+
+  ASSERT (Generator != NULL);
+  ASSERT (CfgMgrProtocol != NULL);
+  ASSERT (RootNode != NULL);
+  ASSERT (PlicAplicInfo != NULL);
+
+  IcNode = NULL;
+
+  // ASL: Scope (\_SB) {}
+  Status = AmlCodeGenScope (SB_SCOPE, RootNode, &ScopeNode);
+  if (EFI_ERROR (Status)) {
+    ASSERT (0);
+    return Status;
+  }
+
+  ASSERT (Uid <= 0xFF);
+  // Write the name of the Interrupt Controller device.
+  CopyMem (AslName, "ICxx", AML_NAME_SEG_SIZE + 1);
+  AslName[AML_NAME_SEG_SIZE - 2] = AsciiFromHex (Uid & 0xFF);

```

Comment:
Shouldn't this look like:

```
  AslName[AML_NAME_SEG_SIZE - 1] = AsciiFromHex (Uid & 0xF);
  if (Uid > 0xF) {
    AslName[AML_NAME_SEG_SIZE - 2] = AsciiFromHex ((Uid >> 4) & 0xF);
  }
```

Reply:
This is better except the condition > 0XF is not required..